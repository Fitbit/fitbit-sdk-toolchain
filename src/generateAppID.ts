import os from 'os';

import fsExtra from 'fs-extra';
import lodash from 'lodash';
import uuid from 'uuid';

export default function generateAppID() {
  const configPath = 'package.json';
  const config = fsExtra.readJSONSync(configPath);
  if (!lodash.isPlainObject(config)) {
    throw new TypeError('Project configuration root must be an object');
  }

  if (config.fitbit === undefined) config.fitbit = {};

  if (!lodash.isPlainObject(config.fitbit)) {
    throw new TypeError(
      "Project configuration 'fitbit' property must be an object",
    );
  }

  config.fitbit.appUUID = uuid.v4();

  fsExtra.writeJSONSync(configPath, config, { spaces: 2, EOL: os.EOL });

  console.log(`Wrote new app ID: ${config.fitbit.appUUID}`);
}
