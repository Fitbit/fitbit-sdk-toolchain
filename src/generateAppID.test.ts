import os from 'os';

import fsExtra from 'fs-extra';
import uuid from 'uuid';

import generateAppID from './generateAppID';

jest.mock('fs-extra');
jest.mock('uuid');

const CONFIG_PATH = 'package.json';
const MOCK_UUID = '8d8daf4a-6d92-4df7-a5d2-2655bc2eadae';

let configWriteSpy: jest.MockInstance<typeof fsExtra.writeJSONSync>;

function mockConfigContent(config: any) {
  jest.spyOn(fsExtra, 'readJSONSync').mockReturnValueOnce(config);
}

beforeEach(() => {
  configWriteSpy = jest.spyOn(fsExtra, 'writeJSONSync');
  jest.spyOn(uuid, 'v4').mockReturnValue(MOCK_UUID);
});

it('throws an error if package.json is not an object', () => {
  mockConfigContent('foo');
  expect(generateAppID).toThrowErrorMatchingSnapshot();
});

it('throws an error if package.json fitbit section is not an object', () => {
  mockConfigContent({ fitbit: false });
  expect(generateAppID).toThrowErrorMatchingSnapshot();
});

it.each([
  ['no fitbit key exists', {}],
  ['a fitbit key exists', { fitbit: {} }],
  ['an existing app ID is present', { fitbit: { appUUID: 'foo' } }],
])('writes a new UUID when %s', (_, configContent) => {
  mockConfigContent(configContent);
  generateAppID();
  expect(configWriteSpy).toBeCalledWith(
    CONFIG_PATH,
    expect.objectContaining({
      fitbit: {
        appUUID: MOCK_UUID,
      },
    }),
    {
      spaces: 2,
      EOL: os.EOL,
    },
  );
});
