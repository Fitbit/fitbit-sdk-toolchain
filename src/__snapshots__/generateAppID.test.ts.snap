// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws an error if package.json fitbit section is not an object 1`] = `"Project configuration 'fitbit' property must be an object"`;

exports[`throws an error if package.json is not an object 1`] = `"Project configuration root must be an object"`;
